### ISSUE TITLE
spack install --overwrite doesn't put build dependencies in build environment

### ISSUE THREAD

**Issue by marcmengel** [LLNL/spack#23428]
Tried to reinstall a package with spack install --overwrite and it failed to find "cmake" which was a build dependency.  

This happens with either concretizer turned on.

### Steps to reproduce the issue

* Make sure you do not have "cmake" in your path.
* fresh install of spack v0.16.1
* Install mysql (if you don't have it already)
* Try to re-install mysql with --overwrite

```console
$ which cmake
/usr/bin/which: no cmake in (...)
$ spack install mysql@5.7.27
...
$ spack find --long mysql
==> 1 installed package
-- linux-scientific7-ivybridge / gcc@4.8.5 ----------------------
xhxn7te mysql@5.7.27

$ spack install --overwrite mysql/xhxn7te    # hash from above
==> The following package specs will be reinstalled:

-- linux-scientific7-x86_64 / gcc@9.3.0 -------------------------
euf24m3 mysql@8.0.19%gcc +client_only~ipo build_type=RelWithDebInfo cxxstd=17
==> Do you want to proceed? [y/N] y
...
==> mysql: Executing phase: 'cmake'
==> Error: ProcessError: cmake: No such file or directory
...
```

So in the plain "spack install" case, it *did* find add cmake to the build environment
but in the --overwrite case it didn't.    

You can work around this by "spack load"-ing the build dependencies before doing
the spack install --overwrite  in the case of cmake, but this doesn't solve all build 
dependency issues.

### Information on your system

$ spack debug report
* **Spack:** 0.16.1
* **Python:** 3.6.8
* **Platform:** linux-scientific7-ivybridge

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output [debug_out.txt](https://github.com/spack/spack/files/6423251/debug_out.txt)

**Comment by sethrj** [comment]
I also have hit this -- today while trying to `spack install --overwrite py-numpy/mno76z`, which has build-time python dependencies:
```
==> py-numpy: Executing phase: 'build'
==> [2021-10-13-14:55:03.472419] '/rnsdhpc/code/spack/opt/spack/apple-clang/python/3tmjst4/bin/python3.8' '-s' 'setup.py' '--no-user-cfg' 'build'
Processing numpy/random/_bounded_integers.pxd.in
Processing numpy/random/_philox.pyx
Traceback (most recent call last):
  File "/private/var/folders/fy/x2xtwh1n7fn0_0q2kk29xkv9vvmbqb/T/s3j/spack-stage/spack-stage-py-numpy-1.21.2-mno76zawnp5wssold5em2ol5jmagzken/spack-src/tools/cythonize.py", line 59, in process_pyx
    import Cython
ModuleNotFoundError: No module named 'Cython'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/private/var/folders/fy/x2xtwh1n7fn0_0q2kk29xkv9vvmbqb/T/s3j/spack-stage/spack-stage-py-numpy-1.21.2-mno76zawnp5wssold5em2ol5jmagzken/spack-src/tools/cythonize.py", line 240, in <module>
    main()
  File "/private/var/folders/fy/x2xtwh1n7fn0_0q2kk29xkv9vvmbqb/T/s3j/spack-stage/spack-stage-py-numpy-1.21.2-mno76zawnp5wssold5em2ol5jmagzken/spack-src/tools/cythonize.py", line 236, in main
    find_process_files(root_dir)
  File "/private/var/folders/fy/x2xtwh1n7fn0_0q2kk29xkv9vvmbqb/T/s3j/spack-stage/spack-stage-py-numpy-1.21.2-mno76zawnp5wssold5em2ol5jmagzken/spack-src/tools/cythonize.py", line 227, in find_process_files
    process(root_dir, fromfile, tofile, function, hash_db)
  File "/private/var/folders/fy/x2xtwh1n7fn0_0q2kk29xkv9vvmbqb/T/s3j/spack-stage/spack-stage-py-numpy-1.21.2-mno76zawnp5wssold5em2ol5jmagzken/spack-src/tools/cythonize.py", line 193, in process
    processor_function(fromfile, tofile)
  File "/private/var/folders/fy/x2xtwh1n7fn0_0q2kk29xkv9vvmbqb/T/s3j/spack-stage/spack-stage-py-numpy-1.21.2-mno76zawnp5wssold5em2ol5jmagzken/spack-src/tools/cythonize.py", line 66, in process_pyx
    raise OSError(msg) from e
OSError: Cython needs to be installed in Python as a module
Running from numpy source directory.
Cythonizing sources
Traceback (most recent call last):
  File "setup.py", line 448, in <module>
    setup_package()
  File "setup.py", line 430, in setup_package
    generate_cython()
  File "setup.py", line 236, in generate_cython
    raise RuntimeError("Running cythonize failed!")
RuntimeError: Running cythonize failed!
```

While trying to reinstall Silo:
```
==> silo: Executing phase: 'autoreconf'
==> Error: RuntimeError: Cannot generate configure: missing dependencies: autoconf, automake, libtool.

Please add the following lines to the package:

    depends_on('autoconf', type='build', when='@4.10.2')
    depends_on('automake', type='build', when='@4.10.2')
    depends_on('libtool', type='build', when='@4.10.2')

Update the version (when='@4.10.2') as needed.

/rnsdhpc/code/spack/lib/spack/spack/build_systems/autotools.py:324, in _require_build_deps:
        321            msg += ("    depends_on('{0}', type='build', when='@{1}')\n"
        322                    .format(dep, spec.version))
        323
  >>    324        msg += "\nUpdate the version (when='@@{0}') as needed.".format(spec.version)
        325        raise RuntimeError(msg)

See build log for details:
  /var/folders/fy/x2xtwh1n7fn0_0q2kk29xkv9vvmbqb/T/s3j/spack-stage/spack-stage-silo-4.10.2-o2brfrtwvxuaoolohqp6imypjcpytlhd/spack-build-out.txt
```

@alalazo @haampie ?

**Comment by sethrj** [comment]
My workaround was to get the full spec (all dependencies and variants), `spack uninstall -f`, and `spack install <full_spec>`... not always possible if you update spack occasionally.

**Comment by adamjstewart** [comment]
@sethrj I think your problem is different and should be fixed by #28346
